zebra-network/src/peer_set/set/tests/prop.rs

```diff
@@ -0,0 +1,135 @@
+};
+use crate::{
+    peer::{LoadTrackedClient, MinimumPeerVersion},
+    peer_set::PeerSet,
+    protocol::external::types::Version,
+};
+
+proptest! {
+    /// Check if discovered outdated peers are immediately dropped by the [`PeerSet`].
+    #[test]
+    fn only_non_outdated_peers_are_accepted(
+        network in any::<Network>(),
+        block_height in any::<block::Height>(),
+        peer_versions in any::<PeerVersions>(),
+    ) {
+        let runtime = zebra_test::init_async();
+
+        let (discovered_peers, mut handles) = peer_versions.mock_peer_discovery();
+        let (mut minimum_peer_version, best_tip_height) =
+            MinimumPeerVersion::with_mock_chain_tip(network);
+
+        best_tip_height
+            .send(Some(block_height))
+            .expect("receiving endpoint lives as long as `minimum_peer_version`");
+
+        let current_minimum_version = minimum_peer_version.current();
+
+        runtime.block_on(async move {
+            let (mut peer_set, _peer_set_guard) = PeerSetBuilder::new()
+                .with_discover(discovered_peers)
+                .with_minimum_peer_version(minimum_peer_version)
+                .build();
+
+            check_if_only_up_to_date_peers_are_live(
+                &mut peer_set,
+                &mut handles,
+                current_minimum_version,
+            )?;
+
+            Ok::<_, TestCaseError>(())
+        })?;
+    }
+
+    /// Check if peers that become outdated after a network upgrade are dropped by the [`PeerSet`].
+    #[test]
+    fn outdated_peers_are_dropped_on_network_upgrade(
+        block_heights in any::<BlockHeightPairAcrossNetworkUpgrades>(),
+        peer_versions in any::<PeerVersions>(),
+    ) {
+        let runtime = zebra_test::init_async();
+
+        let (discovered_peers, mut handles) = peer_versions.mock_peer_discovery();
+        let (mut minimum_peer_version, best_tip_height) =
+            MinimumPeerVersion::with_mock_chain_tip(block_heights.network);
+
+        best_tip_height
+            .send(Some(dbg!(block_heights.before_upgrade)))
+            .expect("receiving endpoint lives as long as `minimum_peer_version`");
+
+        runtime.block_on(async move {
+            let (mut peer_set, _peer_set_guard) = PeerSetBuilder::new()
+                .with_discover(discovered_peers)
+                .with_minimum_peer_version(minimum_peer_version.clone())
+                .build();
+
+            check_if_only_up_to_date_peers_are_live(
+                &mut peer_set,
+                &mut handles,
+                dbg!(minimum_peer_version.current()),
+            )?;
+
+            best_tip_height.send(Some(dbg!(block_heights.after_upgrade)))
+            .expect("receiving endpoint lives as long as `minimum_peer_version`");
+
+            check_if_only_up_to_date_peers_are_live(
+                &mut peer_set,
+                &mut handles,
+                dbg!(minimum_peer_version.current()),
+            )?;
```

Comment:
Nit: Remove `dbg!` macros

```suggestion
        best_tip_height
            .send(Some(block_heights.before_upgrade))
            .expect("receiving endpoint lives as long as `minimum_peer_version`");

        runtime.block_on(async move {
            let (mut peer_set, _peer_set_guard) = PeerSetBuilder::new()
                .with_discover(discovered_peers)
                .with_minimum_peer_version(minimum_peer_version.clone())
                .build();

            check_if_only_up_to_date_peers_are_live(
                &mut peer_set,
                &mut handles,
                minimum_peer_version.current(),
            )?;

            best_tip_height
                .send(Some(block_heights.after_upgrade))
                .expect("receiving endpoint lives as long as `minimum_peer_version`");

            check_if_only_up_to_date_peers_are_live(
                &mut peer_set,
                &mut handles,
                minimum_peer_version.current(),
            )?;
```